### PR TITLE
Optimization pass focused on foam code (saves about 30% of cpu usage I think)

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Foam is crummy at high load rn, both because it runs on a low priority background subsystem, and because it wastes a bit of time.
Let's reduce usage (while speeding up a bunch of other stuff too), and give it more cpu generally.

[Optimizes reagent processing somewhat](https://github.com/tgstation/tgstation/commit/d409bd4afc3c208cd6f00ff406e1e9f78d5ac5ad)

Turns out most of the cost of foam is the reagents it carries, and the varying effects they have
I'm doing my best here to optimize them without touching "user space" too much

That means doing things like prechecking if we're gonna spawn on top of an existing decal (from glitter, flour, etc), and using that same proc to also avoid spawning on unacceptable turfs (I had to convert inheritance to a bitflag system to make this work, but I think that's ok since we want it imparative anyhow)

It's actually nice for code quality too, since it lets me clean up code that was using raw locates and weird var pong.
god I wish I had implied types man

[Optimizes foam spreading in its most accursed aspect, reagent copying](https://github.com/tgstation/tgstation/commit/5cc56a64ad1a22ba7467cb0446b9558560259437)

Holy shit reagent code is a lot.

I'm doing a bunch of small things here. istype in init -> typecache, removing procs that are called once and loop over a list we JUST looped over (ph and the caching for reactions in particular)

I am mainly trying to optimize copy_to here, since that's what foam spams
As a part of this, I removed a pair of update_total and handle_reactions calls that were done on the reagents we are copying FROM

I have no god damn idea why you would want to do that, but if anything is relying on the copy proc modifying the source, then that code deserves to break

Speaking of, I cleaned up handle_reaction's main filter loop a lot, removed a lot of redundant vars and changed it from a full loop w tracker vars to an early exit pattern

This meant using a loop label, which is unfortunate, but this is the fastest method, and it does end up cleaning up the code significantly,
Which is nice

Oh also I made the required_other var function even if there is no atom attached to the reaction, since I don't see why it wouldn't

This last bit is gonna get a bit esoteric so bear with me

Failing calls (which are most of them) to handle_reactions are going to be fastest if they need to check as few reactions as possible

One reagent in a reaction's required list is marked as the "primary", and thus gets to trigger checking it. 
We need all the reagents to react anyhow, so we might as well only check if we have one particular one to avoid double checking

Anyhow, in order to make most calls the fastest, we want these reactions distributed as evenly as possible across all our reagents. 
The current way of doing this is just taking the first reagent in the requirements list and using it, which is not ideal

Instead of that, lets figure out how many reactions each reagent is in, then divy reactions up based off that and the currently divvied reactions

This doubles the reagent index count, and takes the most common reagent, water, from 67 reactions to I think like 22

Does some other general cleaning in reagent code too, etc etc etc

[Fixes runtimes from the forced gravity element being applied more then once](https://github.com/tgstation/tgstation/commit/941d0676114fd455a585f2c65ffc79b81e8438b7)

I feel like this element should take a trait source or something to make them potentially unique, it's too easy to accidentally override one with another

[Removes connect_loc usage in atmos_sensitive, replaces it with direct reg/unreg](https://github.com/tgstation/tgstation/commit/de1c76029d5c49dff152f0ea168b9e6c4a4a04aa)

I only really used it because I liked the componentization, but it costs like 0.2 seconds off init alone which is really stupid, so let's just do this the very slightly harder way

[Micros foam code slightly by inlining a LinkBlockedWithAccess call](https://github.com/tgstation/tgstation/commit/744da3694cd4a85b3bdf44d754de57d7570bdd1c)

This is in the space of like 0.05 seconds kinda save so I can put it back if you'd like, the double loop just felt silly

[Changes how foam processes slightly](https://github.com/tgstation/tgstation/commit/ee5e633e3256fe7df229af71d78424d502459c16)

Rather then treating spreading and processing as separate actions, we do both in sync.
This makes foam fade faster when spreading, which is good cause the whole spread but unclearing foam thing looks silly.
It also avoids the potential bad ending of foam spreading into itself, backwards and forwards. This is better I promise. 

[Bumps fluid priority closer to heavy eaters, moves it off background](https://github.com/tgstation/tgstation/commit/811797f09db7b060f75f15ad06d0ce8982375f47)

Also fixes a bug where foam would travel under public access airlocks.

## Why It's Good For The Game

Saves a lot of cpu just in general, from both init and live.
In theory makes foam faster, tho I'd have to test that on live at highpop to see if I've actually succeeded or not. Guess we'll see.

### CODE REVIEW

**Comment by san7890** [comment]
does this all have to be bundled together? i was giving it a cursory look and the whole TGUI stuff got distracting while I was trying to look for all the changes to fluid code. would also be neater for stuff like tracking this kinda thing in milestones or easy-TM-reverts in case something yonks and we don't lose out on all those other performance gains indefinitely

**Comment by LemonInTheDark** [comment]
> does this all have to be bundled together? i was giving it a cursory look and the whole TGUI stuff got distracting while I was trying to look for all the changes to fluid code. would also be neater for stuff like tracking this kinda thing in milestones or easy-TM-reverts in case something yonks and we don't lose out on all those other performance gains indefinitely
> 
> it's okay if it has to be in a mega-PR but i'd rather more atomicization? it's good work but it's a pain to have to keep scrolling hither and thither

Yeah sorry I uh, I got all focused last night/today and just sort of lumped everything in together.
I could atomize if you'd like, tho I would like this open for now since I'd like to TM and see if things work as I'd like.

If it helps I've got the commits neatly sectioned off, so if it's less of a workload thing and more inspecting different "classes" of changes those should be useful to you.

**Comment by san7890** [comment]
that's fine then, but i think that having more split-out would probably be better for analysis-sakes? it's ultimately whatever though but it's just my personal preference at some point

File: code/__HELPERS/reagents.dm [hunk 1]
```diff
@@ -73,10 +73,10 @@
 /proc/add_chemical_reaction(datum/chemical_reaction/R)
 	if(!GLOB.chemical_reactions_list_reactant_index || !R.required_reagents || !R.required_reagents.len)
 		return
-	var/primary_reagent = R.required_reagents[1]
-	if(!GLOB.chemical_reactions_list_reactant_index[primary_reagent])
-		GLOB.chemical_reactions_list_reactant_index[primary_reagent] = list()
-	GLOB.chemical_reactions_list_reactant_index[primary_reagent] += R
+	var/rand_reagent = pick(R.required_reagents)
```
Comment (jlsnow301) [inline code/__HELPERS/reagents.dm:76]: while you're here, can you fix this single letter var `R` please

Comment (LemonInTheDark) [reply code/__HELPERS/reagents.dm:76]: sure

**Pull request merged**